File: apps/desktop/src/lib/quick-recall/current-frame.ts
// Current-frame ask: the frontend half of "what's on screen right now"
// (round-4 decisions G1–G3).
//
// Hand-mirrored from `crates/capture-types/src/current_frame.rs` — no codegen,
// so this file and that one move together (the Rust serde round-trip test plus
// `bun run check` are the only guards).
//
// The frontend decides NOTHING here. It invokes the capture, renders the chip
// from what came back, and hands the same record straight back to
// `ask_ai_start` / `ask_ai_followup`; the backend re-resolves pixels-vs-OCR-text
// at send time.

import { invoke } from "@tauri-apps/api/core";

export type CurrentFrameCapture = {
  imagePath: string;
  capturedAtUnixMs: number;
  appName: string | null;
  windowTitle: string | null;
  /** Privacy-listed apps that were on screen and got blanked. Named, never hidden. */
  excludedAppNames: string[];
  /** Whether this turn's model can read images. Drives the upfront disclosure. */
  visionSupported: boolean;
  modelLabel: string;
};

/** Take the live screenshot. Throws with a humane reason when Ask AI is unset. */
export function captureCurrentFrame(): Promise<CurrentFrameCapture> {
  return invoke<CurrentFrameCapture>("capture_current_frame");
}

/**
 * Resize the Quick Access window itself — the same window collapses to the bar
 * (G3), never a second one. `null` restores the full launcher.
 */
export function setQuickRecallCollapsed(height: number | null): Promise<void> {
  return invoke<void>("quick_recall_set_collapsed", { height });
}

/** Collapsed heights. Bar alone, and bar plus the detached answer piece. */
export const CURRENT_FRAME_BAR_HEIGHT = 96;
/** Extra height for the non-vision disclosure line, which sits under the bar. */
export const CURRENT_FRAME_DISCLOSURE_HEIGHT = 30;
export const CURRENT_FRAME_ANSWER_HEIGHT = 460;

/** The chip's primary label: what the shot is of. */
export function frameChipLabel(frame: CurrentFrameCapture): string {
  const app = frame.appName?.trim();
  return app && app.length > 0 ? app : "This screen";
}

/**
 * The blanked-apps naming the chip carries. Never a refusal and never silent —
 * an excluded frontmost app still produced a shot, it is just named as missing.
 */
export function frameExclusionNote(frame: CurrentFrameCapture): string | null {
  const names = frame.excludedAppNames.filter((name) => name.trim().length > 0);
  if (names.length === 0) {
    return null;
  }
  if (names.length === 1) {
    return `${names[0]} excluded`;
  }
  if (names.length === 2) {
    return `${names[0]} and ${names[1]} excluded`;
  }
  return `${names.slice(0, -1).join(", ")} and ${names[names.length - 1]} excluded`;
}

/**
 * The upfront non-vision disclosure (G2). Present on the chip BEFORE the user
 * types — the feature stays available, it just sends text instead of pixels.
 */
export function frameVisionNote(frame: CurrentFrameCapture): string | null {
  if (frame.visionSupported) {
    return null;
  }
  const model = frame.modelLabel.trim();
  return `${model.length > 0 ? model : "This model"} can't see images — sending text from this screen`;
}

File: crates/capture-types/src/current_frame.rs
//! The one-shot current-frame capture handed to the frontend and handed back on
//! send (round-4 decisions G1–G3).
//!
//! Hand-mirrored in `apps/desktop/src/lib/quick-recall/current-frame.ts` — there
//! is no codegen, so the round-trip test below plus `bun run check` are the only
//! guards against drift.
//!
//! The frontend never decides anything from this: it renders the chip
//! (`excludedAppNames`, `visionSupported`, `modelLabel`) and hands the whole
//! record back with the question. The backend re-derives pixels-vs-OCR-text at
//! send time.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CurrentFrameCapture {
    /// Absolute path of the JPEG written at invoke time.
    pub image_path: String,
    pub captured_at_unix_ms: i64,
    /// The app owning the frontmost non-excluded window, if any.
    #[serde(default)]
    pub app_name: Option<String>,
    #[serde(default)]
    pub window_title: Option<String>,
    /// Privacy-listed apps that were on screen and got blanked. Named in the
    /// chip ("1Password excluded") — never silently dropped.
    #[serde(default)]
    pub excluded_app_names: Vec<String>,
    /// Whether the model this turn will resolve to can read images. `false`
    /// drives the upfront chip disclosure AND the OCR-text send path.
    pub vision_supported: bool,
    /// The model id the disclosure names, e.g. "llama3.1 can't see images".
    #[serde(default)]
    pub model_label: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn current_frame_capture_round_trips_in_camel_case() {
        let capture = CurrentFrameCapture {
            image_path: "/tmp/mnema-current-frame.jpg".to_string(),
            captured_at_unix_ms: 1_717_000_123_456,
            app_name: Some("Safari".to_string()),
            window_title: Some("Docs".to_string()),
            excluded_app_names: vec!["1Password".to_string()],
            vision_supported: false,
            model_label: "llama3.1".to_string(),
        };

        let json = serde_json::to_value(&capture).expect("serialize");
        assert_eq!(json["imagePath"], "/tmp/mnema-current-frame.jpg");
        assert_eq!(json["capturedAtUnixMs"], 1_717_000_123_456i64);
        assert_eq!(json["appName"], "Safari");
        assert_eq!(json["windowTitle"], "Docs");
        assert_eq!(json["excludedAppNames"][0], "1Password");
        assert_eq!(json["visionSupported"], false);
        assert_eq!(json["modelLabel"], "llama3.1");

        let back: CurrentFrameCapture = serde_json::from_value(json).expect("deserialize");
        assert_eq!(back, capture);
    }

    #[test]
    fn optional_fields_default_when_absent() {
        let back: CurrentFrameCapture = serde_json::from_str(
            r#"{"imagePath":"/tmp/a.jpg","capturedAtUnixMs":1,"visionSupported":true}"#,
        )
        .expect("deserialize");
        assert_eq!(back.app_name, None);
        assert!(back.excluded_app_names.is_empty());
        assert!(back.model_label.is_empty());
    }
}
